Create math-mode

ashok-khanna opened the pull request:
### Brief summary of what the package does

Easy access to 80 math symbols via a Hydra popup menu. Symbol keybindings are carefully thought through to be sensible and mnemonic (as much as possible). Covers most requirements in Math.

### Direct link to the package repository

https://github.com/ashok-khanna/math-mode

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

package-lint giving me the error "This key sequence is reserved (see Key Binding Conventions in the Emacs Lisp manual)". However I want to stick to C-c m and C-c n as the purpose is for easy math shortcuts and thus a more cumbersome keybindings defeats the purpose of the whole package.

### Checklist

<!-- Please confirm with `x`: -->

- [ x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [ x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [ x] My elisp byte-compiles cleanly
- [ x] `M-x checkdoc` is happy with my docstrings
- [ x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented on recipes/math-mode:
```diff
@@ -0,0 +1,3 @@
+(math-mode :repo "ashok-khanna/math-mode"
+                :fetcher github
+                :files ("math-mode.el"))
```
```suggestion
(math-mode :repo "ashok-khanna/math-mode" :fetcher github)
```
This recipe is equivalent and MELPA prefers the default recipe when possible -- it gives you room to grow.

riscy reviewed (commented):
Thanks, this looks useful.  Have you had a chance to compare to some existing packages with similar aims?  E.g. you can [search melpa](https://melpa.org/#/?q=symbols) and look at packages like `char-menu`, `math-symbols`, `xah-math-input`, etc.

> However I want to stick to C-c m and C-c n as the purpose is for easy math shortcuts

I understand the motivation, but this is something we push back against since `C-c <a-z>` are the *only* keybindings that a user is guaranteed to never have taken from them.  One thing I suggest you can do is _recommend_ these keybindings in your `;;; Commentary` instead of imposing them.

One last minor note is that I notice your repository is GPL licensed but the license boilerplate in your file appears to be the MIT license.

ashok-khanna commented:
Hi Chris, many thanks for your time and valuable input. Much appreciated.

I had a look at the packages manually, but thank you so much for this link - it makes searching the Melpa Package much easier.

I did a review of the main packages thus far, they are a bit different and fall into the following categories:

- Autocompletion tools
- Conversion tools from latex to unicode and vice versa
- Shortcut packages (similar to mine), but I couldn't find one that I personally would use.

There were a few relevant packages. One automatically converted English letters to greeks (`xah-math-input`), however I would like to use "a" as "a" and not as alpha as "a" is also a common variable name in maths. `char-menu` may be quite good, however it requires the user to define their own keybindings for each symbol (most of my time was spent choosing a good set of defaults that covered as much math as possible but still mnemonic enough to facilitate easy remembering by the user), and math-symbols seemed quite good, but again wasn't aimed at data entry. I will try and incorporate their ideas around symbol <-> latex conversions at a future date (otherwise one can just pair my mode with theirs).

I have updated the license in the boilerplate to GPL - thanks for that, and also changed the keybindings & added in commentary as you suggested.

Thanks,
Ashok

riscy commented:
Thanks for taking a look at those.  In looking at the bindings you've chosen I can see your rationale, and I think they're a good collection and I see the value.  Another one you may want to take a look at is [company-math](https://melpa.org/#/company-math).

I'm tempted to just hit merge, but I think there are a couple ways we could make this better for the Emacs community.  One of them is the package name.  Since this is a collection of bindings (with an incidental minor mode) I think it's too general to call this `math-mode`.  Would you consider renaming the package to something more specific, like `math-hydras`?  It could help narrow a user's search.

The other is the lack of customizability.  I'd be tempted in this case to drop the minor mode and just let users set up the keybindings into the hydras however they want.  But if you're supporting a minor mode (and thinking of extending in that direction), it may be a good idea to define the keymap using `defcustom` instead of hardcoding it into the mode's definition.

Another customizability concern I have is that -- if users start using your mode and discover they regularly make use of a symbol not on the list (say ⊆), what should they do?  I wouldn't consider this a blocker, but it may be worth it to put all of these bindings into a modifiable data structure, just so you can capture a few more of the long tail use-cases.

ashok-khanna commented:
Many thanks Chris, this was really helpful for a first time package developer like myself and I will use these pointers in future packages I develop.

Math-Hydras sounds great and also makes the connection to the use of hydras in the package.

With regard to customising (which is important as Math is so varying in its use) I guess my question is should we simply leave this as a set of bindings that I share via Reddit or other places for those interested, or should we publish in MELPA - which one do you advise?

If it makes sense to aide discoverability via MELPA, I will make the necessary changes to make it more flexible for users. However I think more now that this should just go into a .emacs as the more I use it, the more I am using it as a global set of key bindings and a 'mode' in particular.

riscy commented:
> I guess my question is should we simply leave this as a set of bindings that I share via Reddit or other places for those interested, or should we publish in MELPA - which one do you advise?

This is a good question!  There are some clear-cut cases where we say "this is configuration more than it's a package" with a few exceptions around things like color themes.  I would mostly lean toward that being the case here, even though you've packed a lot of knowledge into these keybindings.  And yeah -- that's effectively the source of my feedback above. I think I did a good job of beating around the bush. :)

I also have a couple scrappy Emacs packages that I haven't (yet) indexed on MELPA because I feel like they need a bit more time to mature -- and I also personally view putting something on MELPA as acknowledging that I plan to take maintenance somewhat seriously.  So there's that, too.

Maybe the best move here is to let this package live in your configuration for a little bit and see where it goes.  I'd definitely recommend posting it to reddit (or as a gist somewhere).

ashok-khanna commented:
Thanks Chris, will do :)

riscy commented:
Thanks for understanding. :)  Looking forward to seeing more when you get the chance.
